src/libxtp/gaussian_quadrature.cc

```diff
@@ -0,0 +1,225 @@
+    }
+
+    for (Index j = 0; j < _opt.order; ++j) {
+      double newpoint = _quadpoints(j);
+      double newweight = _quadadaptedweights(j);
+      Eigen::VectorXcd denominator1 =
+          (1.0) / (DeltaE + std::complex<double>(0.0, newpoint));
+      Eigen::MatrixXcd Amx = denominator1.asDiagonal() * Imx;
+      Eigen::MatrixXcd Cmx = Imx * (_dielinv_matrices_r[j]);
+      std::complex<double> value1 = (Cmx.cwiseProduct(Amx)).sum();
+      Eigen::VectorXcd denominator2 =
+          (1.0) / (DeltaE + std::complex<double>(0.0, -newpoint));
+      Eigen::MatrixXcd Dmx = denominator2.asDiagonal() * Imx;
+      Eigen::MatrixXcd Emx = Imx * (_dielinv_matrices_r[j].conjugate());
+      std::complex<double> value2 = (Emx.cwiseProduct(Dmx)).sum();
+
+      result += newweight * (value1 + value2);
+    }
+
+    result *= 0.5;
+  } else if (_opt.quadrature_scheme == "modified_legendre") {
+    // This work for integration limits a = 0 b = +infty
+
+    for (Index k = 0; k < lumo; ++k) {
+      DeltaE(k) += std::complex<double>(0.0, 1.0 * eta);
+    }
+    for (Index k = lumo; k < _energies.size(); ++k) {
+      DeltaE(k) += std::complex<double>(0.0, -1.0 * eta);
+    }
+
+    for (Index j = 0; j < _opt.order; ++j) {
+      double exponent = (1.0 + _quadpoints(j)) / (1.0 - _quadpoints(j));
+      double newpoint = std::pow(0.5, exponent);
+      double den =
+          (1.0 - _quadadaptedweights(j)) * (1.0 - _quadadaptedweights(j));
+      double newweight = (2.0 * _quadadaptedweights(j) * 0.5) / den;
+      Eigen::VectorXcd denominator1 =
+          (1.0) / (DeltaE + std::complex<double>(0.0, newpoint));
+      Eigen::MatrixXcd Amx = denominator1.asDiagonal() * Imx;
+      Eigen::MatrixXcd Cmx = Imx * (_dielinv_matrices_r[j]);
+      std::complex<double> value1 = (Cmx.cwiseProduct(Amx)).sum();
+      Eigen::VectorXcd denominator2 =
+          (1.0) / (DeltaE + std::complex<double>(0.0, -newpoint));
+      Eigen::MatrixXcd Dmx = denominator2.asDiagonal() * Imx;
+      Eigen::MatrixXcd Emx = Imx * (_dielinv_matrices_r[j].conjugate());
+      std::complex<double> value2 = (Emx.cwiseProduct(Dmx)).sum();
+
+      result += newweight * (value1 + value2);
+    }
+
+    result *= 0.5;
+
+  } else if (_opt.quadrature_scheme == "hermite") {
+    // This work for integration limits a = -infty b = +infty
+
+    for (Index k = 0; k < lumo; ++k) {
+      DeltaE(k) += std::complex<double>(0.0, 1.0 * eta);
+    }
+    for (Index k = lumo; k < _energies.size(); ++k) {
+      DeltaE(k) += std::complex<double>(0.0, -1.0 * eta);
+    }
+
+    for (Index j = 0; j < _opt.order; ++j) {
+
+      Eigen::VectorXcd denominator =
+          (1.0) / (DeltaE + std::complex<double>(0.0, _quadpoints(j)));
+      Eigen::MatrixXcd Amx = denominator.asDiagonal() * Imx;
+      Eigen::MatrixXcd Cmx = Imx * (_dielinv_matrices_r[j]);
+      std::complex<double> value = (Cmx.cwiseProduct(Amx)).sum();
+      result += _quadadaptedweights(j) * value;
+    }
```

Comment:
this is all basically the same as before. Please refactor but either making the individual quadrature schemes, derived classes of some `quadrature_scheme_base` or so. If you need help let me know.